### PR TITLE
add of queue function

### PR DESCRIPTION
**Pull request by EagleOne33**

### CODE REVIEW

**Comment by guilhermesad** [comment]
@EagleOne33 looks good, thanks! Can you add `@param`s to the documentation as in other methods? Also, I think `uri` param should be before `device_id` so the user doesn't have to type `nil` every time.

**Comment by EagleOne33** [comment]
For now I'm holding it because I have an issue. I keep getting a 503 Service Unavailable error as a response to my call... I originally thought it would be the reason why you didn't implement it in the first place.
I contacted Spotify who told me to ask the question on StackOverflow... [Here it is](https://stackoverflow.com/questions/61992264/spotify-api-queue-method-returning-503-service-unavailable-error)
No answer yet

**Comment by guilhermesad** [comment]
@EagleOne33 that's interesting, so maybe we should remove both doc examples not to have `nil` explicitly.

**Comment by EagleOne33** [comment]
Yes but that's not really the problem unfortunately

**Comment by guilhermesad** [comment]
Yep I understand, just saying that's a nice-to-have

**Comment by EagleOne33** [comment]
I'm out of options to make this work. I also tried [a post there](https://community.spotify.com/t5/Spotify-for-Developers/Queue-function-returning-503-error/m-p/4963373#M281) but no support.

Would you give it a try to tell me if you get the same error?

File: lib/rspotify/player.rb [hunk 1]
```diff
@@ -71,6 +71,20 @@ def play(device_id = nil, params = {})
...
+    #
+    # @example
+    #           player = user.player
+    #           player.queue(nil, "spotify:track:4iV5W9uYEdYUVa79Axb7Rh")
+    def queue(device_id = nil, uri)
+      url = "me/player/queue"
+      url = device_id.nil? ? url : "#{url}?device_id=#{device_id}"
+      param = {"uri": uri}
+
+      User.oauth_put(@user.id, url, param.to_json)
```
Comment (dersam) [inline lib/rspotify/player.rb:85]: This API endpoint is a `POST` not a `PUT`. [reference](https://developer.spotify.com/documentation/web-api/reference-beta/#endpoint-add-to-queue)
```suggestion
      User.oauth_post(@user.id, url, {})
```

File: lib/rspotify/player.rb [hunk 2]
```diff
@@ -71,6 +71,20 @@ def play(device_id = nil, params = {})
...
+    # Add an item to the end of the user’s current playback queue
+    # If `device_id` is not passed, the currently active spotify app will be triggered
+    #
+    # @example
+    #           player = user.player
+    #           player.queue(nil, "spotify:track:4iV5W9uYEdYUVa79Axb7Rh")
+    def queue(device_id = nil, uri)
+      url = "me/player/queue"
+      url = device_id.nil? ? url : "#{url}?device_id=#{device_id}"
+      param = {"uri": uri}
```
Comment (dersam) [inline lib/rspotify/player.rb:83]: The URI needs to be passed as a query string like the device ID.

**Comment by EagleOne33** [comment]
@dersam  I made this change but I have another error: 

```
2.7.0 :007 > player.queue(nil, "spotify:track:4iV5W9uYEdYUVa79Axb7Rh")
Traceback (most recent call last):
        1: from (irb):7
RestClient::BadRequest (400 Bad Request)

```

Is there something else you changed to make it work ?

File: lib/rspotify/player.rb [hunk 3]
```diff
@@ -71,6 +71,19 @@ def play(device_id = nil, params = {})
...
```
Comment (dersam) [inline lib/rspotify/player.rb:83]: This needs to be added to the query string. `param.to_json` adding it to the body won't be understood by Spotify.
```suggestion
```

File: lib/rspotify/player.rb [hunk 4]
```diff
@@ -71,6 +71,19 @@ def play(device_id = nil, params = {})
...
+      User.oauth_post(@user.id, url, param.to_json)
```
Comment (dersam) [inline lib/rspotify/player.rb:84]:
```suggestion
      User.oauth_post(@user.id, url, {})
```

File: lib/rspotify/player.rb [hunk 5]
```diff
@@ -71,6 +71,19 @@ def play(device_id = nil, params = {})
...
 
+    # Add an item to the end of the user’s current playback queue
+    # If `device_id` is not passed, the currently active spotify app will be triggered
+    #
+    # @example
+    #           player = user.player
+    #           player.queue(nil, "spotify:track:4iV5W9uYEdYUVa79Axb7Rh")
+    def queue(device_id = nil, uri)
+      url = "me/player/queue"
+      url = device_id.nil? ? url : "#{url}?device_id=#{device_id}"
```
Comment (dersam) [inline lib/rspotify/player.rb:82]:
```suggestion
      url = device_id.nil? ? url : "#{url}&device_id=#{device_id}"
```

File: lib/rspotify/player.rb [hunk 6]
```diff
@@ -71,6 +71,19 @@ def play(device_id = nil, params = {})
...
     end
 
+    # Add an item to the end of the user’s current playback queue
+    # If `device_id` is not passed, the currently active spotify app will be triggered
+    #
+    # @example
+    #           player = user.player
+    #           player.queue(nil, "spotify:track:4iV5W9uYEdYUVa79Axb7Rh")
+    def queue(device_id = nil, uri)
+      url = "me/player/queue"
```
Comment (dersam) [inline lib/rspotify/player.rb:81]:
```suggestion
      url = "me/player/queue?uri=#{uri}"
```

**Comment by EagleOne33** [comment]
Yes it's fixed now !

**Comment by dersam** [comment]
@EagleOne33 Why did you close this? It hasn't been merged.

**Comment by EagleOne33** [comment]
> @EagleOne33 Why did you close this? It hasn't been merged.

mistake... :(

**Comment by patrickschulze** [comment]
Any update on when this will be merged? Seems like a very useful feature.

**Comment by EagleOne33** [comment]
@guilhermesad I have made the required changes that we talked about at the beginning.
I think it's ready for a merge.
Let me know

Cheers

**Comment by EagleOne33** [comment]
@guilhermesad : possible to merge?

**Comment by guilhermesad** [comment]
@EagleOne33 looks good, thanks!

**Pull request merged**